rfcs/20191127-pip-structure.md

```diff
@@ -0,0 +1,212 @@
+* Jump-to-definition doesn’t work.
+* Quick documentation doesn’t work.
+
+
+## User Benefit
+
+TensorFlow package structure creates difficulties for those who use IDEs.
+Autocomplete, quick documentation and jump-to-definition features often rely on
+module structure matching directory structure. For example, TensorFlow code uses
+`from tensorflow.foo` imports but lives under tensorflow\_core package. Simplifying
+package structure would improve productivity for TensorFlow users.
+
+## Design Proposal
+
+The best way I can think of to fix the autocomplete issues is to make our package structure as clean as possible. In this case, autocomplete will work out of the box.
+
+### Short term: Remove virtual pip package
+
+Primary purpose of keeping the virtual pip package is to workaround circular
+estimator imports. Alternatively, we can resolve this issue by lazy loading
+estimator.
+
+Lazy loading by itself would mean that we no longer have autocomplete for estimator. As a workaround, we can import estimator in the stub *.pyi* file.
+
+Estimator import in root *\_\_init\_\_.py* file:
+```python
+from tensorflow.python.util.lazy_loader import LazyLoader as _LazyLoader
+estimator = _LazyLoader(
+    "estimator", globals(),
+    "tensorflow_estimator.python.estimator.api._v2.estimator")
+setattr(_current_module, "estimator", estimator)
+```
+
+Estimator import in root *\_\_init\_\_.pyi* file:
+```python
+from tensorflow_estimator.python.estimator.api._v2 import estimator as estimator
+```
+In Python 3, we don't even need the *.pyi* file and instead can import estimator
+without lazy loading if `typing.TYPE_CHECKING` is `True`.
+
+After building a pip package with this change all of the following work in PyCharms (both released and EAP) and VS Code:
+
+* jump-to-definition
+* quick documentation
+* autocomplete for `compat.v1.keras`, `compat.v2.keras`
+* autocomplete for keras when using from tensorflow import keras
+* ...basically any import I tested works with autocompletion
+
+To support the TensorFlow Metapackage plans we could add a new pip package that specifies dependencies on tensorflow, tensorflow\_estimator, tensorboard, etc.. Its sole purpose would be to get all dependencies installed.
+
+![alt_text](https://github.com/annarev/community/blob/pip_structure_rfc/rfcs/20191127-pip-structure/new_modular_structure.png "New proposed modular TensorFlow structure.")
+
+### Long term: Import from external package directly
+Short term would fix IDE issues, but the package structure is still not as clean as it could be. We resolve cycles with lazy loading but it would be even better not to have this circular structure at all.
+
+Therefore, I propose that we don’t import external packages into TensorFlow 3.0. Users who want to use estimator, tensorboard summaries or keras could import them separately:
+
+Current code that looks like:
+```python
+import tensorflow as tf
+
+tf.estimator
+tf.keras
+tf.summary
+```
+
+Would be changed to:
```

Comment:
These are API-breaking changes, right? If so, it may be good to spell out the version horizon of this. This will not happen until 3.0, presumably.